File: 动态规划/813最大平均值和的分组/sol.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var largestSumOfAverages = function (nums, k) {
  const len = nums.length;
  nums.sort((a,b)=>a-b);
  console.log(nums);
  const prefix = new Array(len + 1).fill(0);
  for (let i = 1; i <= len; i++) {
    prefix[i] = prefix[i - 1] + nums[i - 1];
  }
  const dp = new Array(len + 1).fill(0).map(() => new Array(k + 1).fill(0));

  for (let i = 1; i <= len; i++) {
    dp[i][1] = prefix[i] / i;
  }
  for (let j = 2; j <= k; j++) {
    for (let i = j; i <= len; i++) {
      for (let x = j - 1; x < i; x++) {
        dp[i][j] = Math.max(dp[i][j], dp[x][j - 1] + (prefix[i] - prefix[x]) / (i - x));
      }
    }
  }
  console.log(dp);
  return dp[len][k];
};

const nums = [-1, 2,-1, 3], k = 3
console.log(largestSumOfAverages(nums, k));

File: 动态规划/813最大平均值和的分组/sol.py
from typing import List
from itertools import accumulate

class Solution:
    def largestSumOfAverages(self, nums: List[int], k: int) -> float:
        n = len(nums)
        # 前缀和
        prefix = list(accumulate(nums, initial=0))
        # 使用dp[i][j]表示nums在区间[0,i-1]被切分成j个子数组的最大平均值和
        dp = [[0.0] * (k + 1) for _ in range(n + 1)]
        for i in range(1, n + 1):
            dp[i][1] = prefix[i] / i
        # 分组个数
        for j in range(2, k + 1):
            for i in range(j, n + 1):
                # 分成[0,x-1] 和[x, i-1]两部分 x>=j-1
                for x in range(j - 1, i):
                    dp[i][j] = max(dp[i][j], dp[x][j - 1] + (prefix[i] - prefix[x]) / (i - x))
        return dp[n][k]


Sol = Solution()
nums = [-1, 2,-1, 3]
k = 3
res = Sol.largestSumOfAverages(nums, k)
print(res)